# Basic Data Processing with Pandas

From Coursera: Intro to Data Science, Week 2  
Patricia Schuster, University of Michigan  
August 2017

# DataFrame Loading, Querying, Indexing

A common workflow:

* Read data into a dataframe, usually from an external file
* Reduce dataframe to only the columns that you need

In [1]:
import pandas as pd
import os

# Loading data

Pandas has built-in support for delimited files such as `.csv` files. We will look at data about olympic medals in the `../data` folder.

In [2]:
df = pd.read_csv('../data/olympics.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

It looks like the first row of data should be the column names, and the first column with country names should be the indices. Use input parameters for `read_csv` to make this happen.

In [18]:
df = pd.read_csv('../data/olympics.csv', index_col = 0, skiprows = 1)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

This data was imported from `https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table`, and the column labels for gold, silver, and medal came through as `01 !`, etc. We want to modify those using `.rename`.

In [23]:
for col in df.columns:
    print(col, col[4:])

№ Summer mmer
01 ! 
02 ! 
03 ! 
Total l
№ Winter nter
01 !.1 .1
02 !.1 .1
03 !.1 .1
Total.1 l.1
№ Games mes
01 !.2 .2
02 !.2 .2
03 !.2 .2
Combined total ined total


In [29]:
for col in df.columns:
    if col[:2] == '01':
        df.rename(columns={col:'Gold'+col[4:]}, inplace=True)
    if col[:2] == '02':
        df.rename(columns={col:'Silver'+col[4:]}, inplace=True)
    if col[:2] == '03':
        df.rename(columns={col:'Bronze'+col[4:]}, inplace=True)

In [30]:
for col in df.columns:
    print(col)

№ Summer
Gold
Silver
Bronze
Total
№ Winter
Gold.1
Silver.1
Bronze.1
Total.1
№ Games
Gold.2
Silver.2
Bronze.2
Combined total


In [31]:
df.head()

№ Summer  Gold  Silver  Bronze  Total  № Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  № Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

# Querying a DataFrame

A Boolean mask is an array where each value is either True or False. It is overlaid on top of a data structure that we're querying, and any cell aligned with a True value will be retained in our final result.

Boolean masks are created by applying operators directly to a pandas DataFrame. 

Create a Boolean mask that shows which countries have received at least one summer gold medal.

In [32]:
df['Gold'] > 0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
Australia (AUS) [AUS] [Z]                        True
Austria (AUT)                                    True
Azerbaijan (AZE)                                 True
Bahamas (BAH)                                    True
Bahrain (BRN)                                   False
Barbados (BAR) [BAR]                            False
Belarus (BLR)                                    True
Belgium (BEL)                                    True
Bermuda (BER)                                   False
Bohemia (BOH) [BOH] [Z]                         False
Botswana (BOT)                                  False
Brazil (BRA)                                     True
British West Indies (BWI) [BWI]                 False
Bulgaria (BUL) [H]          

Now overlay that mask on the DataFrame using the `where` function.

In [34]:
only_gold = df.where(df['Gold'] > 0)
only_gold.head()

№ Summer  Gold  Silver  Bronze  Total  № Winter  \
Afghanistan (AFG)             NaN   NaN     NaN     NaN    NaN       NaN   
Algeria (ALG)                12.0   5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)              23.0  18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                 5.0   1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]       2.0   3.0     4.0     5.0   12.0       0.0   

                         Gold.1  Silver.1  Bronze.1  Total.1  № Games  Gold.2  \
Afghanistan (AFG)           NaN       NaN       NaN      NaN      NaN     NaN   
Algeria (ALG)               0.0       0.0       0.0      0.0     15.0     5.0   
Argentina (ARG)             0.0       0.0       0.0      0.0     41.0    18.0   
Armenia (ARM)               0.0       0.0       0.0      0.0     11.0     1.0   
Australasia (ANZ) [ANZ]     0.0       0.0       0.0      0.0      2.0     3.0   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)             NaN       NaN             NaN  
Algeria (ALG)                 2.0       8.0            15.0  
Argentina (ARG)              24.0      28.0            70.0  
Armenia (ARM)                 2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       4.0       5.0            12.0

Here, all of the countries that did not meet the condition have `NaN` data instead. 

Count how many entries there are (how many countries have a summer gold).

In [39]:
only_gold['Gold'].count()

100

How many countries were there originally?

In [37]:
df['Gold'].count()

147

How can we drop the rows that don't have any data? Use the `drop_na` function. 

In [41]:
only_gold = only_gold.dropna()
only_gold.head()

№ Summer   Gold  Silver  Bronze  Total  № Winter  \
Algeria (ALG)                  12.0    5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)                23.0   18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                   5.0    1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]         2.0    3.0     4.0     5.0   12.0       0.0   
Australia (AUS) [AUS] [Z]      25.0  139.0   152.0   177.0  468.0      18.0   

                           Gold.1  Silver.1  Bronze.1  Total.1  № Games  \
Algeria (ALG)                 0.0       0.0       0.0      0.0     15.0   
Argentina (ARG)               0.0       0.0       0.0      0.0     41.0   
Armenia (ARM)                 0.0       0.0       0.0      0.0     11.0   
Australasia (ANZ) [ANZ]       0.0       0.0       0.0      0.0      2.0   
Australia (AUS) [AUS] [Z]     5.0       3.0       4.0     12.0     43.0   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 5.0       2.0       8.0            15.0  
Argentina (ARG)              18.0      24.0      28.0            70.0  
Armenia (ARM)                 1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]   144.0     155.0     181.0           480.0

This is a more direct way of querying the DataFrame:

In [44]:
df[df['Gold']>0].head()

№ Summer  Gold  Silver  Bronze  Total  № Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   

                           Gold.1  Silver.1  Bronze.1  Total.1  № Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480

Now be more generous. Look for countries that have received a gold in the summer OR winter games (working from the summer and winter columns, not from the total column). 

Note that each Boolean mask has to be encased in parenthesis.

In [51]:
df[(df['Gold'] > 0) | (df['Gold.1'] > 0)].head()

№ Summer  Gold  Silver  Bronze  Total  № Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   

                           Gold.1  Silver.1  Bronze.1  Total.1  № Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480

In [52]:
len(df[(df['Gold'] > 0) | (df['Gold.1'] > 0)])

101

So there is one country who has only won a gold in the winter olympics, never the summer olympics. Which country is it? 

In [54]:
df[(df['Gold'] == 0) & (df['Gold.1'] > 0)]

№ Summer  Gold  Silver  Bronze  Total  № Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  № Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

# Indexing DataFrames

## Basics

In [55]:
df.head()

№ Summer  Gold  Silver  Bronze  Total  № Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  № Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

What if we don't want to index by the name of the country, but rather by the number of gold medals that were won at summer games?

Generate a new column to preserve the country info.

In [58]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

№ Summer  Silver  Bronze  Total  № Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  № Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

It looks like there is an empty row, but actually the index has a name. Also notice that the `Gold` column is gone. 

We can get rid of the index entirely by using the function `reset_index()`, which changes the indices to integers and moves the current index values to a column. 

In [59]:
df = df.reset_index()
df.head()

Gold  № Summer  Silver  Bronze  Total  № Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  № Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

## Multi-level indexing

Pandas has the option to do multi-level indexing. We simply call set index and give a list of columns. This is particularly useful for geographical data which has many layers of geographic regions.

Look at some census data.

In [64]:
df = pd.read_csv('../data/census.csv')
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  

[5 rows x 100 columns]

In [69]:
len(df)

3193

There is a lot of data here. 3193 rows and 100 columns!

In [66]:
df.columns

Index(['SUMLEV', 'REGION', 'DIVISION', 'STATE', 'COUNTY', 'STNAME', 'CTYNAME',
       'CENSUS2010POP', 'ESTIMATESBASE2010', 'POPESTIMATE2010',
       'POPESTIMATE2011', 'POPESTIMATE2012', 'POPESTIMATE2013',
       'POPESTIMATE2014', 'POPESTIMATE2015', 'NPOPCHG_2010', 'NPOPCHG_2011',
       'NPOPCHG_2012', 'NPOPCHG_2013', 'NPOPCHG_2014', 'NPOPCHG_2015',
       'BIRTHS2010', 'BIRTHS2011', 'BIRTHS2012', 'BIRTHS2013', 'BIRTHS2014',
       'BIRTHS2015', 'DEATHS2010', 'DEATHS2011', 'DEATHS2012', 'DEATHS2013',
       'DEATHS2014', 'DEATHS2015', 'NATURALINC2010', 'NATURALINC2011',
       'NATURALINC2012', 'NATURALINC2013', 'NATURALINC2014', 'NATURALINC2015',
       'INTERNATIONALMIG2010', 'INTERNATIONALMIG2011', 'INTERNATIONALMIG2012',
       'INTERNATIONALMIG2013', 'INTERNATIONALMIG2014', 'INTERNATIONALMIG2015',
       'DOMESTICMIG2010', 'DOMESTICMIG2011', 'DOMESTICMIG2012',
       'DOMESTICMIG2013', 'DOMESTICMIG2014', 'DOMESTICMIG2015', 'NETMIG2010',
       'NETMIG2011', 'NETMIG2012', 'NETMI

This contains summaried data on three levels: The whole country, each state, and each county. 

What if we want to see all of the unique values in each column? 

In [63]:
df['SUMLEV'].unique()

array([40, 50], dtype=int64)

Reduce the DataFrame a bit and only keep some of the data. Only look at data where `SUMLEV == 50`. These are the summaries at the state level (remove those at the county level).

In [70]:
df = df[df['SUMLEV']==50]
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        
5          57322              57322            57373     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   
5          1.807375         -1.177622         -1.748766         -2.062535   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   
5         -1.369970     1.859511    -0.848580    -1.402476    -1.577232   

   RNETMIG2015  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  
5    -0.884411  

[5 rows x 100 columns]

Reduce the data to the total number of population estimates and the total number of births. Specify which columns we want to keep. 

In [71]:
columns_to_keep = ['STNAME',
                  'CTYNAME',
                  'BIRTHS2010',
                  'BIRTHS2011',
                  'BIRTHS2012',
                  'BIRTHS2013',
                  'BIRTHS2014',
                  'BIRTHS2015',
                  'POPESTIMATE2010',
                  'POPESTIMATE2011',
                  'POPESTIMATE2012',
                  'POPESTIMATE2013',
                  'POPESTIMATE2014',
                  'POPESTIMATE2015']
df = df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

Set the index of the data to the state and county names. This produces a dual index- first the state name, and then the county name. 

In [72]:
df = df.set_index(['STNAME','CTYNAME'])
df.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

How do we query this DataFrame? We must provide the arguments in order by the order in which we wish to query.

Look at data from one county (our home county!):

In [73]:
df.loc['Michigan','Washtenaw County']

BIRTHS2010            977
BIRTHS2011           3826
BIRTHS2012           3780
BIRTHS2013           3662
BIRTHS2014           3683
BIRTHS2015           3709
POPESTIMATE2010    345563
POPESTIMATE2011    349048
POPESTIMATE2012    351213
POPESTIMATE2013    354289
POPESTIMATE2014    357029
POPESTIMATE2015    358880
Name: (Michigan, Washtenaw County), dtype: int64

Look at data across two counties. Provide both pairs of indices. 

In [74]:
df.loc[[('Michigan','Washtenaw County'),('Michigan','Wayne County')]]

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME   CTYNAME                                                            
Michigan Washtenaw County         977        3826        3780        3662   
         Wayne County            5918       23819       23270       23377   

                           BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME   CTYNAME                                                     
Michigan Washtenaw County        3683        3709           345563   
         Wayne County           23607       23586          1815199   

                           POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME   CTYNAME                                                               
Michigan Washtenaw County           349048           351213           354289   
         Wayne County              1801273          1792514          1775713   

                           POPESTIMATE2014  POPESTIMATE2015  
STNAME   CTYNAME                                             
Michigan Washtenaw County           357029           358880  
         Wayne County              1766008          1759335

This is an example of hierarchical labels by rows. You can also do hierarchical labels by columns. 

## Challenge question

Reindex the purchase records DataFrame to be indexed hierarchically, first by store, then by person. Name these indexes 'Location' and 'Name'. Then add a new entry to it with the value of:

Name: 'Kevyn', Item Purchased: 'Kitty Food', Cost: 3.00, Location: 'Store 2'.

In [87]:
purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])

df.head()

Cost Item Purchased   Name
Store 1  22.5       Dog Food  Chris
Store 1   2.5   Kitty Litter  Kevyn
Store 2   5.0      Bird Seed  Vinod

First turn the current indices- the store names- into a column.

In [88]:
df = df.reset_index()
df = df.rename(columns = {'index':'Location'})
df.head()

Location  Cost Item Purchased   Name
0  Store 1  22.5       Dog Food  Chris
1  Store 1   2.5   Kitty Litter  Kevyn
2  Store 2   5.0      Bird Seed  Vinod

In [89]:
df = df.set_index(['Location','Name'])
df.head()

Cost Item Purchased
Location Name                      
Store 1  Chris  22.5       Dog Food
         Kevyn   2.5   Kitty Litter
Store 2  Vinod   5.0      Bird Seed

In [92]:
df = df.append(pd.Series({'Cost':3.00,'Item Purchased':'Kitty Food'}, name=('Store 2','Kevyn')))
df

Cost Item Purchased
Location Name                      
Store 1  Chris  22.5       Dog Food
         Kevyn   2.5   Kitty Litter
Store 2  Vinod   5.0      Bird Seed
         Kevyn   3.0     Kitty Food

# Handling Missing Values

In other fields, it is common to fill missing values with `99`. Pandas has the ability to fill it with `NaN`. 

Look at the file `../data/log.csv`.

In [2]:
df = pd.read_csv('../data/log.csv')
df

time    user          video  playback position paused  volume
0   1469974424  cheryl     intro.html                  5  False    10.0
1   1469974454  cheryl     intro.html                  6    NaN     NaN
2   1469974544  cheryl     intro.html                  9    NaN     NaN
3   1469974574  cheryl     intro.html                 10    NaN     NaN
4   1469977514     bob     intro.html                  1    NaN     NaN
5   1469977544     bob     intro.html                  1    NaN     NaN
6   1469977574     bob     intro.html                  1    NaN     NaN
7   1469977604     bob     intro.html                  1    NaN     NaN
8   1469974604  cheryl     intro.html                 11    NaN     NaN
9   1469974694  cheryl     intro.html                 14    NaN     NaN
10  1469974724  cheryl     intro.html                 15    NaN     NaN
11  1469974454     sue  advanced.html                 24    NaN     NaN
12  1469974524     sue  advanced.html                 25    NaN     NaN
13  1469974424     sue  advanced.html                 23  False    10.0
14  1469974554     sue  advanced.html                 26    NaN     NaN
15  1469974624     sue  advanced.html                 27    NaN     NaN
16  1469974654     sue  advanced.html                 28    NaN     5.0
17  1469974724     sue  advanced.html                 29    NaN     NaN
18  1469974484  cheryl     intro.html                  7    NaN     NaN
19  1469974514  cheryl     intro.html                  8    NaN     NaN
20  1469974754     sue  advanced.html                 30    NaN     NaN
21  1469974824     sue  advanced.html                 31    NaN     NaN
22  1469974854     sue  advanced.html                 32    NaN     NaN
23  1469974924     sue  advanced.html                 33    NaN     NaN
24  1469977424     bob     intro.html                  1   True    10.0
25  1469977454     bob     intro.html                  1    NaN     NaN
26  1469977484     bob     intro.html                  1    NaN     NaN
27  1469977634     bob     intro.html                  1    NaN     NaN
28  1469977664     bob     intro.html                  1    NaN     NaN
29  1469974634  cheryl     intro.html                 12    NaN     NaN
30  1469974664  cheryl     intro.html                 13    NaN     NaN
31  1469977694     bob     intro.html                  1    NaN     NaN
32  1469977724     bob     intro.html                  1    NaN     NaN

We can use the pandas function `fillna` to manage the `NaN` values. Use the method `ffill` for forward filling (fill data based on a previous row). 

In [4]:
df = df.set_index('time')
df = df.sort_index()
df

user          video  playback position paused  volume
time                                                               
1469974424  cheryl     intro.html                  5  False    10.0
1469974424     sue  advanced.html                 23  False    10.0
1469974454  cheryl     intro.html                  6    NaN     NaN
1469974454     sue  advanced.html                 24    NaN     NaN
1469974484  cheryl     intro.html                  7    NaN     NaN
1469974514  cheryl     intro.html                  8    NaN     NaN
1469974524     sue  advanced.html                 25    NaN     NaN
1469974544  cheryl     intro.html                  9    NaN     NaN
1469974554     sue  advanced.html                 26    NaN     NaN
1469974574  cheryl     intro.html                 10    NaN     NaN
1469974604  cheryl     intro.html                 11    NaN     NaN
1469974624     sue  advanced.html                 27    NaN     NaN
1469974634  cheryl     intro.html                 12    NaN     NaN
1469974654     sue  advanced.html                 28    NaN     5.0
1469974664  cheryl     intro.html                 13    NaN     NaN
1469974694  cheryl     intro.html                 14    NaN     NaN
1469974724  cheryl     intro.html                 15    NaN     NaN
1469974724     sue  advanced.html                 29    NaN     NaN
1469974754     sue  advanced.html                 30    NaN     NaN
1469974824     sue  advanced.html                 31    NaN     NaN
1469974854     sue  advanced.html                 32    NaN     NaN
1469974924     sue  advanced.html                 33    NaN     NaN
1469977424     bob     intro.html                  1   True    10.0
1469977454     bob     intro.html                  1    NaN     NaN
1469977484     bob     intro.html                  1    NaN     NaN
1469977514     bob     intro.html                  1    NaN     NaN
1469977544     bob     intro.html                  1    NaN     NaN
1469977574     bob     intro.html                  1    NaN     NaN
1469977604     bob     intro.html                  1    NaN     NaN
1469977634     bob     intro.html                  1    NaN     NaN
1469977664     bob     intro.html                  1    NaN     NaN
1469977694     bob     intro.html                  1    NaN     NaN
1469977724     bob     intro.html                  1    NaN     NaN

Indices are not unique, and several users can use the system at the same time. To manage this, use multi-level indexing of time and user name. 

In [6]:
df = df.reset_index()
df = df.set_index(['time','user'])
df

index          video  playback position paused  volume
time       user                                                          
1469974424 cheryl      0     intro.html                  5  False    10.0
           sue         1  advanced.html                 23  False    10.0
1469974454 cheryl      2     intro.html                  6    NaN     NaN
           sue         3  advanced.html                 24    NaN     NaN
1469974484 cheryl      4     intro.html                  7    NaN     NaN
1469974514 cheryl      5     intro.html                  8    NaN     NaN
1469974524 sue         6  advanced.html                 25    NaN     NaN
1469974544 cheryl      7     intro.html                  9    NaN     NaN
1469974554 sue         8  advanced.html                 26    NaN     NaN
1469974574 cheryl      9     intro.html                 10    NaN     NaN
1469974604 cheryl     10     intro.html                 11    NaN     NaN
1469974624 sue        11  advanced.html                 27    NaN     NaN
1469974634 cheryl     12     intro.html                 12    NaN     NaN
1469974654 sue        13  advanced.html                 28    NaN     5.0
1469974664 cheryl     14     intro.html                 13    NaN     NaN
1469974694 cheryl     15     intro.html                 14    NaN     NaN
1469974724 cheryl     16     intro.html                 15    NaN     NaN
           sue        17  advanced.html                 29    NaN     NaN
1469974754 sue        18  advanced.html                 30    NaN     NaN
1469974824 sue        19  advanced.html                 31    NaN     NaN
1469974854 sue        20  advanced.html                 32    NaN     NaN
1469974924 sue        21  advanced.html                 33    NaN     NaN
1469977424 bob        22     intro.html                  1   True    10.0
1469977454 bob        23     intro.html                  1    NaN     NaN
1469977484 bob        24     intro.html                  1    NaN     NaN
1469977514 bob        25     intro.html                  1    NaN     NaN
1469977544 bob        26     intro.html                  1    NaN     NaN
1469977574 bob        27     intro.html                  1    NaN     NaN
1469977604 bob        28     intro.html                  1    NaN     NaN
1469977634 bob        29     intro.html                  1    NaN     NaN
1469977664 bob        30     intro.html                  1    NaN     NaN
1469977694 bob        31     intro.html                  1    NaN     NaN
1469977724 bob        32     intro.html                  1    NaN     NaN

A lot of the pandas packages are written to handle missing values. For instance, if you use `numpy.mean`, it will exclude any missing values. 